Draggable: Add Getting Started (#8276)
DevExpress/devextreme-documentation: concepts/05 UI Components/Draggable/05 Getting Started with Draggable/15 Handle Events.md

> This example specifies an [onDragStart](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onDragStart) handler to update card overlapping as follows:

Reviewer: ```suggestion
This example specifies an [onDragStart](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onDragStart) handler to update card overlapping:
```

        element.style.zIndex = z.toString();
        z++;
    }

    function handleDragStart(e: DxDraggableTypes.DragStartEvent) {
        changeZIndex(e.element);
    }

    </script>

##### React

    <!-- tab: App.tsx -->
    import React, { JSX, useState, useCallback } from 'react';
    import { Draggable, type DraggableTypes } from 'devextreme-react/draggable';

    function App(): JSX.Element {
        const [, setZ] = useState(1);

        const changeZIndex = useCallback((element: HTMLElement) => {
            setZ((prevValue) => {
                element.style.zIndex = prevValue.toString();
                return prevValue + 1;
            });
        }, []);

        const handleDragStart = useCallback((e: DraggableTypes.DragStartEvent) => {
            changeZIndex(e.element);
        }, []);

        return (
            <div class="board">
                <Draggable ...
                    onDragStart={handleDragStart}
                />

                {/* ... */}
            </div>
        );
    }

---

This example also specifies handlers for four common events (not specific to Draggable):

- `click`
- `dxdragenter`
- `dxdragleave`
- `dxdrop`

---

##### jQuery

> To attach handlers to common events, call [Draggable.on()](/Documentation/ApiReference/UI_Components/dxDraggable/Methods/#onevents) as follows:

Reviewer: ```suggestion
To attach handlers to common events, call [Draggable.on()](/Documentation/ApiReference/UI_Components/dxDraggable/Methods/#onevents):
```

    <!-- tab: index.js -->
    $(function() {
        $("#note-1").dxDraggable({
            group: 'cards',

            // ...
        }).on({
            "click": handleClick,
            "dxdragenter": handleDragEnter,
            "dxdragleave": handleDragStop,
            "dxdrop": handleDragStop,
        });

        // ...

        function handleClick(e) {
            changeZIndex($(e.currentTarget));
        }

        function handleDragEnter(e) {
            e.target.classList.add('overlapped');
        }

        function handleDragStop(e) {
            e.target.classList.remove('overlapped');
        }
    });

##### Angular

> To attach handlers to common events, call the [on()](/Documentation/ApiReference/Common/Utils/events/#onelement_eventName_handler) utility method within [Draggable.onInitialized()](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onInitialized) as follows:

Reviewer: ```suggestion
To attach handlers to common events, call the [on()](/Documentation/ApiReference/Common/Utils/events/#onelement_eventName_handler) utility method within [Draggable.onInitialized()](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onInitialized):
```

            group="cards"
            (onInitialized)="handleInit($event)"
        ></dx-draggable>

        <!-- ... -->
    </div>

    <!-- tab: app.component.ts -->
    import { type DxDraggableTypes } from 'devextreme-angular/ui/draggable';
    import { EventObject } from 'devextreme/common/core/events';
    import { on } from "devextreme/events";

    // ...
    export class AppComponent {
        handleClick(e: EventObject) {
            this.changeZIndex(e.currentTarget as HTMLElement);
        }

        handleDragEnter(e: EventObject) {
            const target: HTMLElement = e.target as HTMLElement;

            target.classList.add('overlapped');
        }

        handleDragStop(e: EventObject) {
            const target: HTMLElement = e.target as HTMLElement;

            target.classList.remove('overlapped');
        }

        handleInit(e: DxDraggableTypes.InitializedEvent) {
            on(e.element!, 'click', this.handleClick.bind(this));

            on(e.element!, 'dxdragenter', this.handleDragEnter);

            on(e.element!, 'dxdragleave', this.handleDragStop);
            on(e.element!, 'dxdrop', this.handleDragStop);
        }
    }

##### Vue

> To attach handlers to common events, call the [on()](/Documentation/ApiReference/Common/Utils/events/#onelement_eventName_handler) utility method within [Draggable.onInitialized()](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onInitialized) as follows:

Reviewer: ```suggestion
To attach handlers to common events, call the [on()](/Documentation/ApiReference/Common/Utils/events/#onelement_eventName_handler) utility method within [Draggable.onInitialized()](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onInitialized):
```

    <template>
        <div class="board">
            <DxDraggable ...
                group="cards"
                @initialized="handleInit"
            />

            <!-- ... -->
        </div>
    </template>

    <script setup lang="ts">
    import { DxDraggable, type DxDraggableTypes } from "devextreme-vue/draggable";
    import { on } from "devextreme/events";
    import { type EventObject } from "devextreme/common/core/events";

    function handleClick(e: EventObject) {
        changeZIndex(e.currentTarget as HTMLElement);
    }

    function handleDragEnter(e: EventObject) {
        const target: HTMLElement = e.target as HTMLElement;

        target.classList.add('overlapped');
    }

    function handleDragStop(e: EventObject) {
        const target: HTMLElement = e.target as HTMLElement;

        target.classList.remove('overlapped');
    }

    function handleInit(e: DxDraggableTypes.InitializedEvent) {
        on(e.element!, "click", handleClick);

        on(e.element!, "dxdragenter", handleDragEnter);

        on(e.element!, "dxdragleave", handleDragStop);
        on(e.element!, "dxdrop", handleDragStop);
    }

    </script>

##### React

> To attach handlers to common events, call the [on()](/Documentation/ApiReference/Common/Utils/events/#onelement_eventName_handler) utility method within [Draggable.onInitialized()](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onInitialized) as follows:

Reviewer: ```suggestion
To attach handlers to common events, call the [on()](/Documentation/ApiReference/Common/Utils/events/#onelement_eventName_handler) utility method within [Draggable.onInitialized()](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onInitialized):
```

---

##### jQuery

> This example specifies an [onDragStart](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onDragStart) handler to update note overlapping:

Reviewer 2: I'm not sure what you mean by "update note overlapping". I imagine notes that overlap each other when a user drags them around. But how or what do you update? The code seems to suggest that you put the note being dragged to the top of Z-index.

```suggestion
This example specifies an [onDragStart](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#onDragStart) handler to update the note's Z-index (bring to front):
```

Author: Yeah, incrementing the z-index value of the active note is exactly what happens, thank you!

- `dxdragenter`
- `dxdragleave`
- `dxdrop`

To attach handlers to common events, call [Draggable.on()](/Documentation/ApiReference/UI_Components/dxDraggable/Methods/#onevents):

    <!-- tab: index.js -->
    $(function() {
        $("#note-1").dxDraggable({
            group: 'notes',

            // ...
        }).on({
            "click": handleClick,
            "dxdragenter": handleDragEnter,
            "dxdragleave": handleDragStop,
            "dxdrop": handleDragStop,
        });

        // ...

        function handleClick(e) {
            changeZIndex($(e.currentTarget));
        }

        function handleDragEnter(e) {
            e.target.classList.add('overlapped');
        }

        function handleDragStop(e) {
            e.target.classList.remove('overlapped');
        }
    });

    <!-- tab: index.css -->
    .note.overlapped {
        outline: 1px dashed var(--dx-color-primary);
    }

[note] Specify an identical [group](/Documentation/ApiReference/UI_Components/dxDraggable/Configuration/#group) value for all Draggable instances to ensure the components interact with each other (in `dxdragenter`, `dxdragleave`, and `dxdrop` handlers).

##### Angular

> This example specifies the following event handlers to update note overlapping:

Reviewer 2: ```suggestion
This example specifies the following event handlers to update the note's Z-index (bring to front while dragging):
```

> This example specifies the following event handlers to update the note's Z-index (bring to front while dragging):

Reviewer: ```suggestion
This example specifies the following event handlers to update the note's Z-index (bring it to the front while dragging):
```

        handleClick(): void {
            this.updateZIndex();
        }
    }

    <!-- tab: app.component.html -->
    <div class="boundary-text">Dragging Boundary</div>
        <div class="board">
            @for (note of notes; track note) {
            <note ...
                [isOverlapped]="overlappedId === note.id"
                [(zIndex)]="zIndex"
                [startOverlap]="startOverlap"
                [stopOverlap]="stopOverlap"
            ></note>
            }
        </div>
    <div class="boundary-text">Dragging Boundary</div>

    <!-- tab: app.component.ts -->
    import { Component } from '@angular/core';

    // ...
    export class AppComponent {
        // ...

        zIndex: number = 0;
        overlappedId: string | null = null;

        startOverlap: (id: string) => void = (id) => {
            this.overlappedId = id;
        }

        stopOverlap: () => void = () => {
            this.overlappedId = null;
        }

        handleDragEnd: () => void = () => {
            this.stopOverlap();
        }
    }

    <!-- tab: app.component.scss -->
    .note.overlapped {
        outline: 1px dashed var(--dx-color-primary);
    }

##### Vue

> This example specifies the following event handlers to update note overlapping:

Reviewer 2: ```suggestion
This example specifies the following event handlers to update the note's Z-index (bring to front while dragging):
```

> This example specifies the following event handlers to update the note's Z-index (bring to front while dragging):

Reviewer: ```suggestion
This example specifies the following event handlers to update the note's Z-index (bring it to the front while dragging):
```

            stopOverlap();
            overlappedComponentId.value = null;
        }
    }

    function handleDragEnd() {
        stopOverlap();
    }

    function handleClick() {
        updateZIndex();
    }
    </script>

    <!-- tab: App.vue -->
    <template>
        <div class="boundary-text">Dragging Boundary</div>
        <div class="board">
            <template v-for="note in notes">
            <Note ...
                :is-overlapped="overlappedId === note.id"
                v-model:z-index="zIndex"
                :start-overlap="startOverlap"
                :stop-overlap="stopOverlap"
            />
            </template>
        </div>
        <div class="boundary-text">Dragging Boundary</div>
    </template>

    <script setup lang="ts">
    import { ref } from "vue";
    // ...

    const overlappedId = ref<string | null>(null);
    const zIndex = ref<number>(0);

    function startOverlap(id: string) {
        overlappedId.value = id;
    }

    function stopOverlap() {
        overlappedId.value = null;
    }
    </script>

    <style scoped>
    .note.overlapped {
        outline: 1px dashed var(--dx-color-primary);
    }

    </style>

##### React

> This example specifies the following event handlers to update note overlapping:

Reviewer 2: ```suggestion
This example specifies the following event handlers to update the note's Z-index (bring to front while dragging):
```
